### ISSUE TITLE
Failure to decode reads overhanging reference end.

### ISSUE THREAD

**Issue by jkbonfield** [divonlan/genozip#12]
From the Illumina Platinum Genomes file for NA12878, a snippet:

```
@SQ	SN:chrM	LN:16571
@RG	ID:NA12878	SM:NA12878
HSQ1004:134:C0D8DACXX:1:1107:20540:135446	101	chrM	16472	37	101M	=	227	-16144	GGGGGTAGCTAAAGTGAACTGTATCCGACATCTGGTTCCTACTTCAGGGTCATAAAGCCTAAATAGCCCACACGTTCCCCTTAAATAAGACATCACGATGG	@C@FFDDFHHFFHIHIIHHEI9FBFHGGGGGGIIJG<FHIHHIIIIJJJIIGJIHGGJIJJJFGHHHHEDFFDDEABDDDDDDCDEEDDDDDACCCB2?B<	RG:Z:NA12878	XT:A:U	NM:i:2	XN:i:1	SM:i:37	AM:i:37	X0:i:1	X1:i:0	XM:i:2	XO:i:0	XG:i:0	MD:Z:49C50C0
```

This extends from 16472 to 16572 inclusive due to the 101M cigar, but the reference ends at 16571.

The file encodes fine without warnings, but decode then fails.  It would be preferable for the encode to fail or (better) for the decode to succeed, even if it's editing cigar to 100M1S.

**Comment by divonlan** [comment]
Hi James,

Thank you so much for reporting this.

This issue won't replicate on my side, also the code is designed to wrap around chromosomes at the end of the reference (with circular chromosomes in mind), but obviously there is some issue.

Could you please tell me the exact command that you ran for compressing and decompressing, the reference file that you used (if you compressed with a reference file and its public), and the error message you see?

Thanks!!

-divon

**Comment by jkbonfield** [comment]
Hi,

The exact commands were:

Encode:
```
$ ../genozip/genozip -f -e hg19.ref.genozip chrM.sam
genozip chrM.sam : Done (0 seconds, SAM compression ratio: 0.1)
```

[ As an aside, can you confirm that this is the best way to maximum lossless compression?  I am aware of `--optimize`, but the manual indicates this is lossy. ]

Decode:
```
$ ../genozip/genocat -e hg19.ref.genozip chrM.sam.genozip
@SQ	SN:chrM	LN:16571
@RG	ID:NA12878	SM:NA12878


Regions set for chrom 90 "chrM" [1-16571] according to range.is_set (format- "first_pos-last_pos (len)")
In the neighborhood of about 10000 bp around pos=16572
No data: pos=16572 is outside of [first_pos=1 - last_pos=16571]


Error in sam_reconstruct_seq:163: Error in sam_reconstruct_seq: while reconstructing line 3 (vb_i=1: last_txt_line=3 num_lines=1): reference is not set: chrom=90 "chrM" pos=16572 range=[1-16571] (cigar=101M seq_start_pos=16472 ref_consumed=100 seq_consumed=100)
```

I get the same error using genounzip too.

The hg19ref.genozip was created by running `./genozip --make-reference`.  I don't know where I orignally obtained my copy of hg19.fa though.  The input SAM was culled from the Platinum Genomes file at ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR194/ERR194147/NA12878_S1.bam

Annoyingly this file has minimal SAM headers, but experience with chrM vs M made me believe hg19.fa as opposed to GRCh37, plus the fact that chrM is 16571 and not 16569 bases.  (I really wish such high profile organisations could get basic data provenance issues correct!)

**Comment by divonlan** [comment]
Hi James, thanks again for reporting this and thanks for your patience. This issue has now been fixed, and released in version 12.0.34.

Regarding the best lossless compression, in addition to using --reference, you can also increase the block size with e.g. -B2048 (in MB, 2048 is the maximum), at the expense of RAM consumption and slower subsetting with eg genocat --regions. You can see the block size set by default and also get some insights into the compression with genocat --stats or (even more detailed) --STATS.
